src/combinatorics/quiver.lean

```diff
@@ -0,0 +1,89 @@
+/-
+Copyright (c) 2021 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn
+-/
+import tactic
+/-!
+# Quivers
+
+This module defines quivers. A quiver `G` on a type `V` of vertices assigns to every
+pair `a b : V` of vertices a type `G a b`, thought of as edges from `a` to `b`. This
+is a very permissive notion of graph.
+-/
+
+-- We use the same universe order as in category theory.
+universes v u
+
+/-- A quiver `G` on a type `V` of vertices assigns to every pair `a b : V` of vertices
+    a type `G a b`, thought of as edges from `a` to `b`. -/
+def quiver (V : Type u) := V → V → Type v
```

Comment:
Any reason not generalize to `Sort`?
```suggestion
/-- A quiver `G` on a type `V` of vertices assigns to every pair `a b : V` of vertices
    a type `G a b`, thought of as edges from `a` to `b`. -/
def quiver (V : Type u) := V → V → Sort v
```
In particular, this makes the `V → V → Prop` of `simple_graph` a special case of this.

Reply:
There's a bit of a conceptual difference between a `quiver` and a `simple_graph` though, which is that `V → V → Prop` for a `simple_graph` isn't recording edge sets exactly.  It's true that you can think of a simple graph as being a directed graph with a pair of directed edges for each edge, but then the directed graph's type of all edges map's 2:1 to the simple graph's type of all edges.

That said, I'd also like to see this be `Sort v` if possible.

Reply:
Would you mind making this be a structure?
```lean
structure quiver (V : Type u) :=
(edges : V → V → Sort v)
```
If `G` is a quiver, I'm a bit more comfortable having the API be `G.edges i j` rather than `G i j`.  I don't have a use case in mind, but this would allow a more refined `quiver` type to `extend` this one.

Reply:
If you use the terminology quiver, edges should be called arrows.

Reply:
We tried at least two times to make the hom-sets in `category_theory` use `Sort*` instead of `Type*`, but failed miserably. I can imagine similar issues here... there are probably remnants of frustrations in the zulip logs if you are looking for inspiration :grinning: .

Reply:
structure & arrow: done.

We can generalize to `Sort*`, but definitions like `subquiver` and `quiver.sum` will somehow assume we're not in `Prop`. So far this doesn't look like a problem, but it might cause problems later.

Reply:
Hmm, as far as I can see, the only problem with `Sort *` is that we'll sometimes need to assume `G : quiver.{v+1} V` instead of `G : quiver V`. Maybe this is a small price to pay if it lets us define things like connectivity uniformly.

It would be nice to have some abbreviation like `foo V := quiver.{v+1} V`, but this doesn't seem to work reliably with dot notation.